resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/IFile.java

```diff
@@ -438,8 +428,14 @@ public default void create(byte[] content, int createFlags, IProgressMonitor mon
 	public default void write(byte[] content, boolean force, boolean derived, boolean keepHistory,
 			IProgressMonitor monitor) throws CoreException {
 		Objects.requireNonNull(content);
-		create(content, (force ? IResource.FORCE : 0) | (derived ? IResource.DERIVED : 0)
-				| (keepHistory ? IResource.KEEP_HISTORY : 0) | IResource.REPLACE, monitor);
+		if (exists()) {
+			if (derived != isDerived()) {
+				setDerived(derived, null);
+			}
+			setContents(content, force, keepHistory, monitor);
```

Comment:
Usually, methods in resource are supposed to be atomic (ie grouped in a single operation), here those 2 methods would lead to 2 operations side by side but not grouped.
What you can try to remediate that would be something like
```java
if (derived == isDerived()) {
	setContents(...);
} else {
	SubMonitor progress = SubMonitor.convert(monitor, 100).checkCanceled();
	// need to group operation
	final ISchedulingRule groupRule = new MultiRule[] { workspace.getRuleFactory().derivedRule(this), workspace.getRuleFactory().modifyRule(this) };
	try {
		workspace.prepareOperation(rule, progress.split(1)); // locks
		workspace.beginOperation(true);
		setDerived(derived, progress.split(1));
		setContents(content, force, keepHistory, progress.split(98))
	} catch (OperationCanceledException e) {
		workspace.getWorkManager().operationCanceled();
		throw e;
	} finally {
		progress.done();
		workspace.endOperation(rule, true); // this is extremely important of workspace is locked
	}
```

Reply:
API-wise it's totally fine. See nested comment about the workspace operation should be atomic.

Reply:
Thanks for the suggestion! I have tried your implementation and it would work but can not be done in default method of IFile due to references to internal methods. but based on the idea i think i have found a easier solution with less overhead. please review again. Junit also tests for atomic operations now.

Reply:
>  in default method of IFile due to references to internal methods. 

There is nothing mandating to create a default method here. IFile is annotated as "noextends/noimplements", so we can add new interface methods in here and have the implementation in File.

Reply:
So far the therory. In practice there are other implementations even within the IDE (PDE's ContributionDataFile) :-(

Reply:
While it's an annoyance, I think keeping the resource code as maintainable as posible is more important than keeping e4 tools alive. IMO, it's best to go for a regular interface method and impl in File, and to adapt PDE then.

Reply:
sure, that would be good. But i have no idea why it is like that in PDE and how it could be solved. So as long as we can implement a default its  much easier to proceed.